torch/csrc/jit/passes/annotate_effects.cpp

```diff
@@ -0,0 +1,184 @@
+ *
+ * This pass annotates effectful operations (such as ones that mutate existing
+ * values) to prevent subsequent passes from re-ordering ops in a way that
+ * changes the meaning of the program.
+ *
+ * It does this by adding primitive memory fence ops around nodes that use
+ * mutable values. The memory fences input and output a "World" token, which
+ * expresses effects explicitly in the IR and forces all fenced nodes to be
+ * linearized.
+ */
+class AnnotateEffectsImpl {
+ public:
+  void annotateEffects(Graph* g) {
+    // Generate the first world token
+    const auto tokenGenerator = g->create(prim::Constant);
+    g->block()->prependNode(tokenGenerator);
+    auto curToken = tokenGenerator->output()->setType(WorldType::get());
+
+    visitBlock(g->block(), curToken);
+  }
+
+ private:
+  Value* visitBlock(Block* block, Value* curToken) {
+    for (auto* node : block->nodes()) {
+      curToken = visitNode(node, curToken);
+    }
+    return curToken;
+  }
+
+  // If a node uses a mutable variable (or mutates a previously constant
+  // variable), create a memory fence around the node.
+  //
+  // Returns the last world token emitted for subsequent memory fences to use.
+  Value* visitNode(Node* node, Value* curToken) {
+    if (node->kind() == prim::If) {
+      JIT_ASSERT(node->blocks().size() == 2);
+
+      auto trueBlock = node->blocks().at(0);
+      auto falseBlock = node->blocks().at(1);
+
+      auto trueToken = visitBlock(trueBlock, curToken);
+      auto falseToken = visitBlock(falseBlock, curToken);
+
+      // If any branch has a mutating op, this node has to output a world token
+      if (trueToken != curToken || falseToken != curToken) {
+        trueBlock->registerOutput(trueToken);
+        falseBlock->registerOutput(falseToken);
+
+        return node->addOutput()->setType(WorldType::get());
+      }
+      return curToken;
+    }
+
+    if (node->kind() == prim::Loop) {
+      JIT_ASSERT(node->blocks().size() == 1);
+      auto block = node->blocks().at(0);
+      if (!shouldAddFence(block)) {
+        // Bail out early if there's no mutable variables used inside
+        return curToken;
+      }
+
+      // Register the world token as a loop carried dependency
+      auto beginLoopToken = block->addInput()->setType(WorldType::get());
+      auto endLoopToken = visitBlock(block, beginLoopToken);
+      block->registerOutput(endLoopToken);
+
+      JIT_ASSERT(endLoopToken != beginLoopToken);
+
+      // Thread the world token through the loop node
+      node->addInput(curToken);
+      return node->addOutput()->setType(WorldType::get());
+    }
+
+    JIT_ASSERT(node->blocks().size() == 0);
+
+    if (shouldAddFence(node)) {
+      return addFenceForNode(node, curToken);
+    }
+
+    return curToken;
+  }
+
+  bool shouldAddFence(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (mutableValues_.count(input) != 0) {
+        return true;
```

Comment:
Oh, so every node that uses a mutable value will need this special node? That doesn't seem right. It should be enough to replace the value at the points where it is mutated, but all reads shouldn't need to do anything later.

Reply:
We need to guard every use of the mutable value if we want to prevent all forms of dependence-violating reordering. As a small example:
```
foo = []
foo.append(5)
x = len(foo)
foo.append(6)
```
Any reordering of instructions here will change the meaning of the program. If we only fenced the `append`s, then only ordering of the appends would be linearized, and the following optimization would be allowed:
```
foo = []
foo.append(5)
foo.append(6)
x = len(foo)
```
It's true that non-mutating uses before the first and after the last mutation could be safe to re-order. I was thinking I'd add something to relax that constraint later.